File: examples/yolov3_coco.cpp
// Tencent is pleased to support the open source community by making ncnn available.
//
// Copyright (C) 2018 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except
// in compliance with the License. You may obtain a copy of the License at
//
// https://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.

#include "net.h"
#include "benchmark.h"

#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <stdio.h>
#include <vector>

struct Object
{
    cv::Rect_<float> rect;
    int label;
    float prob;
};

static int detect_yolov3(const char * paramFile, const char * binFile, const cv::Mat& bgr, std::vector<Object>& objects)
{
    ncnn::Net yolov3;

    yolov3.opt.use_vulkan_compute = true;

    // original pretrained model from https://github.com/eric612/MobileNet-YOLO
    // param : https://drive.google.com/open?id=1V9oKHP6G6XvXZqhZbzNKL6FI_clRWdC-
    // bin : https://drive.google.com/open?id=1DBcuFCr-856z3FRQznWL_S5h-Aj3RawA
    // the ncnn model https://github.com/nihui/ncnn-assets/tree/master/models
    // yolov3.load_param("mobilenetv2_yolov3.param");
    // yolov3.load_model("mobilenetv2_yolov3.bin");
    yolov3.load_param(paramFile);
    yolov3.load_model(binFile);
    printf("Model loaded.\n");

    const int target_size = 416;

    int img_w = bgr.cols;
    int img_h = bgr.rows;

    ncnn::Mat in = ncnn::Mat::from_pixels_resize(bgr.data, ncnn::Mat::PIXEL_BGR, bgr.cols, bgr.rows, target_size, target_size);

    /*const float mean_vals[3] = {127.5f, 127.5f, 127.5f};
    const float norm_vals[3] = {0.007843f, 0.007843f, 0.007843f};*/
    const float mean_vals[3] = {0.f, 0.f, 0.f};
    const float norm_vals[3] = {1/255.f, 1/255.f, 1/255.f};
    in.substract_mean_normalize(mean_vals, norm_vals);

    double start = ncnn::get_current_time();
    for (int c = 0; c < 20; ++c) {

        ncnn::Extractor ex = yolov3.create_extractor();

        int r = ex.input("data", in);

        ncnn::Mat out;
        ex.extract("output", out);

        printf("out w/h/c: %d %d %d\n", out.w, out.h, out.c);
        objects.clear();
        for (int i = 0; i < out.h; i++)
        {
            const float* values = out.row(i);

            Object object;
            object.label = values[0];
            object.prob = values[1];
            object.rect.x = values[2] * img_w;
            object.rect.y = values[3] * img_h;
            object.rect.width = values[4] * img_w - object.rect.x;
            object.rect.height = values[5] * img_h - object.rect.y;
            printf("label: %f, prob: %f\n", values[0], values[1]);

            objects.push_back(object);
        }
    }
    double end = ncnn::get_current_time();
    double time = end - start;
    printf("Average inference time:%7.2f \n", time/20);

    return 0;
}

static void draw_objects(const cv::Mat& bgr, const std::vector<Object>& objects)
{
	/*
    static const char* class_names[] = {"background",
                                        "aeroplane", "bicycle", "bird", "boat",
                                        "bottle", "bus", "car", "cat", "chair",
                                        "cow", "diningtable", "dog", "horse",
                                        "motorbike", "person", "pottedplant",
                                        "sheep", "sofa", "train", "tvmonitor"
                                       };
	*/
    static const char* class_names[] = {"person",
										"bicycle",
										"car",
										"motorbike",
										"aeroplane",
										"bus",
										"train",
										"truck",
										"boat",
										"traffic light",
										"fire hydrant",
										"stop sign",
										"parking meter",
										"bench",
										"bird",
										"cat",
										"dog",
										"horse",
										"sheep",
										"cow",
										"elephant",
										"bear",
										"zebra",
										"giraffe",
										"backpack",
										"umbrella",
										"handbag",
										"tie",
										"suitcase",
										"frisbee",
										"skis",
										"snowboard",
										"sports ball",
										"kite",
										"baseball bat",
										"baseball glove",
										"skateboard",
										"surfboard",
										"tennis racket",
										"bottle",
										"wine glass",
										"cup",
										"fork",
										"knife",
										"spoon",
										"bowl",
										"banana",
										"apple",
										"sandwich",
										"orange",
										"broccoli",
										"carrot",
										"hot dog",
										"pizza",
										"donut",
										"cake",
										"chair",
										"sofa",
										"pottedplant",
										"bed",
										"diningtable",
										"toilet",
										"tvmonitor",
										"laptop",
										"mouse",
										"remote",
										"keyboard",
										"cell phone",
										"microwave",
										"oven",
										"toaster",
										"sink",
										"refrigerator",
										"book",
										"clock",
										"vase",
										"scissors",
										"teddy bear",
										"hair drier",
										"toothbrush"};

    cv::Mat image = bgr.clone();

    for (size_t i = 0; i < objects.size(); i++)
    {
        const Object& obj = objects[i];

        fprintf(stderr, "%d = %.5f at %.2f %.2f %.2f x %.2f\n", obj.label, obj.prob,
                obj.rect.x, obj.rect.y, obj.rect.width, obj.rect.height);

        cv::rectangle(image, obj.rect, cv::Scalar(255, 0, 0));

        char text[256];
        sprintf(text, "%s %.1f%%", class_names[obj.label-1], obj.prob * 100);

        int baseLine = 0;
        cv::Size label_size = cv::getTextSize(text, cv::FONT_HERSHEY_SIMPLEX, 0.5, 1, &baseLine);

        int x = obj.rect.x;
        int y = obj.rect.y - label_size.height - baseLine;
        if (y < 0)
            y = 0;
        if (x + label_size.width > image.cols)
            x = image.cols - label_size.width;

        cv::rectangle(image, cv::Rect(cv::Point(x, y), cv::Size(label_size.width, label_size.height + baseLine)),
                      cv::Scalar(255, 255, 255), -1);

        cv::putText(image, text, cv::Point(x, y + label_size.height),
                    cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 0, 0));
    }

    cv::imshow("image", image);
    cv::waitKey(0);
}

int main(int argc, char** argv)
{
    if (argc != 4)
    {
        fprintf(stderr, "Usage: %s [imagepath] [param file path] [bin file path]\n", argv[0]);
        return -1;
    }

    const char* imagepath = argv[1];

    cv::Mat m = cv::imread(imagepath, 1);
    if (m.empty())
    {
        fprintf(stderr, "cv::imread %s failed\n", imagepath);
        return -1;
    }

    std::vector<Object> objects;
    detect_yolov3(argv[2], argv[3], m, objects);

    draw_objects(m, objects);

    return 0;
}

File: python/examples/yolov3_coco.py
import numpy as np
import cv2
import time
import os
import argparse
import ncnn

class_names = ["person",
    "bicycle",
    "car",
    "motorbike",
    "aeroplane",
    "bus",
    "train",
    "truck",
    "boat",
    "traffic light",
    "fire hydrant",
    "stop sign",
    "parking meter",
    "bench",
    "bird",
    "cat",
    "dog",
    "horse",
    "sheep",
    "cow",
    "elephant",
    "bear",
    "zebra",
    "giraffe",
    "backpack",
    "umbrella",
    "handbag",
    "tie",
    "suitcase",
    "frisbee",
    "skis",
    "snowboard",
    "sports ball",
    "kite",
    "baseball bat",
    "baseball glove",
    "skateboard",
    "surfboard",
    "tennis racket",
    "bottle",
    "wine glass",
    "cup",
    "fork",
    "knife",
    "spoon",
    "bowl",
    "banana",
    "apple",
    "sandwich",
    "orange",
    "broccoli",
    "carrot",
    "hot dog",
    "pizza",
    "donut",
    "cake",
    "chair",
    "sofa",
    "pottedplant",
    "bed",
    "diningtable",
    "toilet",
    "tvmonitor",
    "laptop",
    "mouse",
    "remote",
    "keyboard",
    "cell phone",
    "microwave",
    "oven",
    "toaster",
    "sink",
    "refrigerator",
    "book",
    "clock",
    "vase",
    "scissors",
    "teddy bear",
    "hair drier",
    "toothbrush"]

image = cv2.imread(r'C:\Users\DY\Cache\cv_test_video\WIN_6.mp4_20210307_162930.998.png', 1)
#image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
#image_resized = cv2.resize(image, (416, 416), interpolation=cv2.INTER_LINEAR)
data = ncnn.Mat.from_pixels_resize(image.data, ncnn.Mat.PixelType.PIXEL_BGR, 1280, 720, 416, 416)
data.substract_mean_normalize([0.0,0.0,0.0],[1/255.0,1/255.0,1/255.0])

net = ncnn.Net()
net.opt.use_vulkan_compute = True
ret = net.load_param(r'C:\Users\DY\Cache\ncnn\yolov3.param')
if ret != 0:
    print('Failed to load param file')

ret = net.load_model(r'C:\Users\DY\Cache\ncnn\yolov3.bin')
if ret != 0:
    print('Failed to load bin file.')

ex = net.create_extractor()

ret = ex.input('data', data)
if ret != 0:
    print('Failed to run input.')

ret, out = ex.extract('output')

for i in range(0, out.h*6, 6):
    print(f'class: {class_names[int(out[i])-1]}, prob: {out[i+1]}')
    
def parser():
    parser = argparse.ArgumentParser(description="YOLOv3 Object Detection")
    parser.add_argument("--input", type=str, default="",
                        help="image source. It can be a single image, a"
                        "txt with paths to them, or a folder. Image valid"
                        " formats are jpg, jpeg or png.")
    parser.add_argument("--batch_size", default=1, type=int,
                        help="number of images to be processed at the same time")
    parser.add_argument("--bin_file", default="yolov3.param",
                        help="yolov3 .bin file path")
    parser.add_argument("--dont_show", action='store_true',
                        help="windown inference display. For headless systems")
    parser.add_argument("--ext_output", action='store_true',
                        help="display bbox coordinates of detected objects")
    parser.add_argument("--save_labels", action='store_true',
                        help="save detections bbox for each image in yolo format")
    parser.add_argument("--param_file", default="yolov3.param",
                        help="path to param file")
    parser.add_argument("--thresh", type=float, default=.25,
                        help="remove detections with lower confidence")
    return parser.parse_args()

def check_arguments_errors(args):
    assert 0 < args.thresh < 1, "Threshold should be a float between zero and one (non-inclusive)"
    if not os.path.exists(args.param_file):
        raise(ValueError("Invalid param path {}".format(os.path.abspath(args.param_file))))
    if not os.path.exists(args.bin_file):
        raise(ValueError("Invalid bin file path {}".format(os.path.abspath(args.bin_file))))
    if args.input and not os.path.exists(args.input):
        raise(ValueError("Invalid image path {}".format(os.path.abspath(args.input))))

def check_batch_shape(images, batch_size):
    """
        Image sizes should be the same width and height
    """
    shapes = [image.shape for image in images]
    if len(set(shapes)) > 1:
        raise ValueError("Images don't have same shape")
    if len(shapes) > batch_size:
        raise ValueError("Batch size higher than number of images")
    return shapes[0]


def load_images(images_path):
    """
    If image path is given, return it directly
    For txt file, read it and return each line as image path
    In other case, it's a folder, return a list with names of each
    jpg, jpeg and png file
    """
    input_path_extension = images_path.split('.')[-1]
    if input_path_extension in ['jpg', 'jpeg', 'png']:
        return [images_path]
    elif input_path_extension == "txt":
        with open(images_path, "r") as f:
            return f.read().splitlines()
    else:
        return glob.glob(os.path.join(images_path, "*.jpg")) + \
            glob.glob(os.path.join(images_path, "*.png")) + \
            glob.glob(os.path.join(images_path, "*.jpeg"))

'''
def prepare_batch(images, network, channels=3):
    width = darknet.network_width(network)
    height = darknet.network_height(network)

    darknet_images = []
    for image in images:
        image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
        image_resized = cv2.resize(image_rgb, (width, height),
                                   interpolation=cv2.INTER_LINEAR)
        custom_image = image_resized.transpose(2, 0, 1)
        darknet_images.append(custom_image)

    batch_array = np.concatenate(darknet_images, axis=0)
    batch_array = np.ascontiguousarray(batch_array.flat, dtype=np.float32)/255.0
    darknet_images = batch_array.ctypes.data_as(darknet.POINTER(darknet.c_float))
    return darknet.IMAGE(width, height, channels, darknet_images)
'''

net_width = 416
net_height = 416

def image_detection(image_path, network, ex, class_names, class_colors, thresh):
    image = cv2.imread(image_path)
    image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    image_resized = cv2.resize(image_rgb, (net_width, net_height),
                               interpolation=cv2.INTER_LINEAR)

    ex.input("data", ncnn.Mat(image_resized))
    ret, out = ex.extract("output")
    if ret != 0:
        print('Failed to detect objects.')
        return None, None

    '''
    for i in range(out.h):
        const float* values = out.row(i);

        Object object;
        object.label = values[0];
        object.prob = values[1];
        object.rect.x = values[2] * img_w;
        object.rect.y = values[3] * img_h;
        object.rect.width = values[4] * img_w - object.rect.x;
        object.rect.height = values[5] * img_h - object.rect.y;
        printf("label: %f, prob: %f\n", values[0], values[1]);

    image = darknet.draw_boxes(detections, image_resized, class_colors)
    return image_resized, detections
    '''

'''
def batch_detection(network, images, class_names, class_colors,
                    thresh=0.25, hier_thresh=.5, nms=.45, batch_size=4):
    image_height, image_width, _ = check_batch_shape(images, batch_size)
    darknet_images = prepare_batch(images, network)
    batch_detections = darknet.network_predict_batch(network, darknet_images, batch_size, image_width,
                                                     image_height, thresh, hier_thresh, None, 0, 0)
    batch_predictions = []
    for idx in range(batch_size):
        num = batch_detections[idx].num
        detections = batch_detections[idx].dets
        if nms:
            darknet.do_nms_obj(detections, num, len(class_names), nms)
        predictions = darknet.remove_negatives(detections, class_names, num)
        images[idx] = darknet.draw_boxes(predictions, images[idx], class_colors)
        batch_predictions.append(predictions)
    darknet.free_batch_detections(batch_detections, batch_size)
    return images, batch_predictions
'''
'''
def image_classification(image, network, class_names):
    width = darknet.network_width(network)
    height = darknet.network_height(network)
    image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    image_resized = cv2.resize(image_rgb, (width, height),
                                interpolation=cv2.INTER_LINEAR)
    darknet_image = darknet.make_image(width, height, 3)
    darknet.copy_image_from_bytes(darknet_image, image_resized.tobytes())
    detections = darknet.predict_image(network, darknet_image)
    predictions = [(name, detections[idx]) for idx, name in enumerate(class_names)]
    darknet.free_image(darknet_image)
    return sorted(predictions, key=lambda x: -x[1])
'''
'''
def convert2relative(image, bbox):
    """
    YOLO format use relative coordinates for annotation
    """
    x, y, w, h = bbox
    height, width, _ = image.shape
    return x/width, y/height, w/width, h/height

def save_annotations(name, image, detections, class_names):
    """
    Files saved with image_name.txt and relative coordinates
    """
    file_name = name.split(".")[:-1][0] + ".txt"
    with open(file_name, "w") as f:
        for label, confidence, bbox in detections:
            x, y, w, h = convert2relative(image, bbox)
            label = class_names.index(label)
            f.write("{} {:.4f} {:.4f} {:.4f} {:.4f} {:.4f}\n".format(label, x, y, w, h, float(confidence)))
'''
'''
def batch_detection_example():
    args = parser()
    check_arguments_errors(args)
    batch_size = 3
    random.seed(3)  # deterministic bbox colors
    network, class_names, class_colors = darknet.load_network(
        args.config_file,
        args.data_file,
        args.weights,
        batch_size=batch_size
    )
    image_names = ['data/horses.jpg', 'data/horses.jpg', 'data/eagle.jpg']
    images = [cv2.imread(image) for image in image_names]
    images, detections,  = batch_detection(network, images, class_names,
                                           class_colors, batch_size=batch_size)
    for name, image in zip(image_names, images):
        cv2.imwrite(name.replace("data/", ""), image)
    print(detections)
'''

def main():
    args = parser()
    check_arguments_errors(args)

    net = ncnn.Net()
    ret = net.load_param(args.param_file)
    if ret != 0:
        print(f'Failed to load param file {args.param_file}.')
        return

    ret = net.load_model(args.bin_file)
    if ret != 0:
        print(f'Failed to load bin file {args.bin_file}.')
        return

    ex = net.create_extractor()

    images = load_images(args.input)

    index = 0
    while True:
        # loop asking for new image paths if no list is given
        if args.input:
            if index >= len(images):
                break
            image_path = images[index]
        else:
            image_name = input("Enter Image Path: ")
        prev_time = time.time()
        image, detections = image_detection(
            image_path, net, ex, class_names, class_colors, args.thresh
            )
        #if args.save_labels:
        #    save_annotations(image_name, image, detections, class_names)
        #print_detections(detections, args.ext_output)
        now = time.time()
        print(f"Inference time: {now-prev_time}s")
        if not args.dont_show:
            cv2.imshow('Result', image)
            if cv2.waitKey() & 0xFF == ord('q'):
                break
        index += 1

if __name__ == "__main__":
    # unconmment next line for an example of batch processing
    # batch_detection_example()
    #main()
    pass
